ui/src/logs/actions/index.ts

```diff
@@ -51,12 +85,73 @@ export const setSource = (source: Source): SetSourceAction => ({
   },
 })
 
-export const setNamespace = (namespace: Namespace): SetNamespaceAction => ({
-  type: ActionTypes.SetNamespace,
-  payload: {
-    namespace,
-  },
-})
+export const executeHistogramQuery = () => async (
+  dispatch,
+  getState: GetState
+) => {
+  const state = getState()
+  const queryConfig = getDeep<QueryConfig | null>(
+    state,
+    'logs.histogramQueryConfig',
+    null
+  )
+  const timeRange = getDeep<TimeRange | null>(state, 'logs.timeRange', null)
+  const namespace = getDeep<Namespace | null>(
+    state,
+    'logs.currentNamespace',
+    null
+  )
+  const proxyLink = getDeep<string | null>(
+    state,
+    'logs.currentSource.links.proxy',
+    null
+  )
+
+  if (queryConfig && timeRange && namespace && proxyLink) {
+    const query = buildQuery(timeRange, queryConfig)
+    const response = await executeQueryAsync(proxyLink, namespace, query)
+
+    dispatch({
+      type: ActionTypes.SetHistogramData,
+      payload: {
+        data: [{response}],
+      },
+    })
+  }
+}
+
+export const setHistogramQueryConfig = () => async (
+  dispatch,
+  getState: GetState
+) => {
+  const state = getState()
+  const namespace = getDeep<Namespace | null>(
+    state,
+    'logs.currentNamespace',
+    null
+  )
+  const timeRange = getDeep<TimeRange | null>(state, 'logs.timeRange', null)
+
+  if (timeRange && namespace) {
+    const queryConfig = buildHistogramQueryConfig(namespace, timeRange)
+
+    dispatch({
+      type: ActionTypes.SetHistogramQueryConfig,
+      payload: {queryConfig},
+    })
+
+    dispatch(executeHistogramQuery())
+  }
+}
+
+export const setNamespace = (namespace: Namespace) => async dispatch => {
+  dispatch({
```

Comment:
what do you think about adopting the pattern we've used elsewhere for this (and my comments above) where, for example here, you would have a synchronous action creator called `setNamespace` that generates this action object, and then to name the thunk `setNamespaceAsync`, which would dispatch the synchronous action?